### ISSUE TITLE
No resource type found for:v1#Status

### ISSUE THREAD

**Issue by bframke** [fabric8io/kubernetes-client#2244]
After we finished developing the first version of our Watcher with the Quarkus Kubernetes Client we had it run for several days. After some time this Message started to print out in the Logs. 

```
2020-05-25 07:25:05,983 WARN  [io.fab.kub.cli.dsl.int.WatchConnectionManager] (OkHttp https://172.20.0.1/...) Exec Failure: io.fabric8.kubernetes.client.KubernetesClientException: An error has occurred.
	at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:248)
	at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:167)
	at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:152)
	at io.fabric8.kubernetes.client.dsl.internal.WatchHTTPManager.readWatchEvent(WatchHTTPManager.java:290)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onMessage(WatchConnectionManager.java:229)
	at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:322)
	at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)
	at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)
	at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:273)
	at okhttp3.internal.ws.RealWebSocket$1.onResponse(RealWebSocket.java:209)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:174)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:497)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
Caused by: com.fasterxml.jackson.databind.JsonMappingException: No resource type found for:v1#Status
 at [Source: (BufferedInputStream); line: 1, column: 176] (through reference chain: io.fabric8.kubernetes.api.model.WatchEvent["object"])
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.fromObjectNode(KubernetesDeserializer.java:107)
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:83)
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:41)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:288)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:151)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4218)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3259)
	at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:246)
	... 16 more
```

We are using our own custom resource, but this part of Fabric8 we never did touch.

**Comment by rohanKanojia** [comment]
Could you please share code/sample project which we can use to reproduce this error?

**Comment by bframke** [comment]
I will build an example for it and can't really tell. Its this Extension from Quarkus https://github.com/quarkusio/quarkus/tree/master/extensions/kubernetes-client which implements, if the pom.xml are correct, the latest version of it.

**Comment by bframke** [comment]
Okay, I've remade our client we use so that the problem happens with same application environment. The problem happend after 24 hours, not sure if it happens earlier because we can't get the old logs out of the kubernetes anymore.

https://github.com/bframke/quarkus-examples/tree/master/kubernetes-client

If you want to run it locally you need to use docker-compose to setup the kafka and have a kubernetes config which is available and set KUBERNETES_NAMESPACE so the application starts correctly. The CRD is also in the Project and should be deployed into the cluster which is mentioned in the kube config.

If you want to deploy it on the cluster build a docker image with `./gradlew docker` and then use the yamls to deploy it correctly, do not forget to change the image in the `client.yaml`. You also need a Kafka deployed to the System, we use Strimzi for it and you need then to set KUBERNETES_CLIENT_KAFKA_BROKER_LIST in the client.yaml with the according connection to the bootstrap Kafka from Strimzi.

**Comment by bframke** [comment]
Quarkus Kubernets Client uses Fabric8 4.9.0

**Comment by rohanKanojia** [comment]
Would appreciate if you could provide some easy way to reproduce.

**Comment by bframke** [comment]
I try my best to find an easier way. I let our application run for 23h with 2 custom resources on it and now it prints out the error again every second.

**Comment by bframke** [comment]
Okay, I've updated the Project. https://github.com/bframke/quarkus-examples/tree/master/kubernetes-client Stripped it down to a bare minimum, so far to reproduce the error: start the client, deploy CRD, deploy one or two CRs and wait for some hours and after a while the WatchHttpManager tries to read v1#Status and fails to find anything thats named that way and prints out the error every second.

**Comment by bframke** [comment]
Also, as soon as the error happens and you restart the Application after around 2 Minutes it happening again.

**Comment by bframke** [comment]
Okay, regarding the problem after an intense through the model library and what happens between the quarkus native build and these objects I came to this result:

The quarkus native build does remove the Class Status which is part of the Fabric8 Model. With that fixed the application now shutsdown because of an too old resource version: 7424497 (7443937)

**Comment by bframke** [comment]
```
io.fabric8.kubernetes.client.KubernetesClientException: too old resource version: 7424448 (7456839)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onMessage(WatchConnectionManager.java:259)
	at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:322)
	at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)
	at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)
	at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:273)
	at okhttp3.internal.ws.RealWebSocket$1.onResponse(RealWebSocket.java:209)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:174)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.lang.Thread.run(Thread.java:834)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:497)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
```

A resource with this version does not even exist on the whole cluster

**Comment by bframke** [comment]
Okay, now we are using the InformerAPI with the ReflectorWatcher from you and so far there are no problems.

**Comment by dtschan** [comment]
> The quarkus native build does remove the Class Status which is part of the Fabric8 Model.

How did you fix the missing Status class? We are hitting the same problem.

**Comment by bframke** [comment]
You could add it to the reflect-config.json, then quarkus will not remove it.

**Comment by metacosm** [comment]
@iocanel @geoand any idea about what's going on here?

**Comment by geoand** [comment]
Doesn't remind me of anything

**Comment by iocanel** [comment]
Same

**Comment by ruromero** [comment]
We are having the same error after some time. The controller becomes unresponsive and has to be restarted:

```
Caused by: com.fasterxml.jackson.databind.JsonMappingException: No resource type found for:v1#Status
 at [Source: (BufferedInputStream); line: 1, column: 183] (through reference chain: io.fabric8.kubernetes.api.model.WatchEvent["object"])
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.fromObjectNode(KubernetesDeserializer.java:85)
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:57)
	at io.fabric8.kubernetes.internal.KubernetesDeserializer.deserialize(KubernetesDeserializer.java:46)
	at com.fasterxml.jackson.databind.deser.impl.MethodProperty.deserializeAndSet(MethodProperty.java:129)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.vanillaDeserialize(BeanDeserializer.java:293)
	at com.fasterxml.jackson.databind.deser.BeanDeserializer.deserialize(BeanDeserializer.java:156)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4526)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3513)
	at io.fabric8.kubernetes.client.utils.Serialization.unmarshal(Serialization.java:243)
	... 16 more
```

Any hint?

All status and spec custom resources are registered in the reflection-config.json
Using quarkus 1.12.0.Final

**Comment by metacosm** [comment]
@ruromero do you have a project that we could use to replicate the issue?

**Comment by robpblake** [comment]
@metacosm Just to add some extra information for you (will work on seeing if we can get a repeater for you)

- We are running a Quarkus-native compiled Operator. @ruromero has listed the version of Quarkus above.
- We have 3 CRDs with a status and we have ensured that both the CRD class and the status class  are listed in `reflection-config.json`
- The issues appears to manifest ~1hr after deployment. Killing the pod brings the Operator back to functionality again for another ~1hr where we then have to repeat the process.

As a test, we have re-deployed the Operator in JVM mode rather than native-compiled mode and we are seeing the following errors in the log after ~1hr:

```
2021-03-09 12:13:48,503 WARN  [org.kie.baa.ccp.con.DecisionEventSource] (OkHttp https://172.30.0.1/...) Received error for watch, will try to reconnect.: io.fabric8.kubernetes.client.WatcherException: too old resource version: 145558586 (146899663)
	at io.fabric8.kubernetes.client.dsl.internal.WatchConnectionManager$1.onMessage(WatchConnectionManager.java:224)
	at okhttp3.internal.ws.RealWebSocket.onReadMessage(RealWebSocket.java:322)
	at okhttp3.internal.ws.WebSocketReader.readMessageFrame(WebSocketReader.java:219)
	at okhttp3.internal.ws.WebSocketReader.processNextFrame(WebSocketReader.java:105)
	at okhttp3.internal.ws.RealWebSocket.loopReader(RealWebSocket.java:273)
	at okhttp3.internal.ws.RealWebSocket$1.onResponse(RealWebSocket.java:209)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:174)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: too old resource version: 145558586 (146899663)
	... 11 more
```

This last error may be a red herring, but I thought it was useful to mention that it appears to happen almost exactly 1hr after deployment. In JVM mode, things appear to recover OK when the error is received.

Any other data we can get you, please let me know.

cheers,

Rob

**Comment by ruromero** [comment]
An update here after some debugging. The error is caused in native build when receiving a Status event which for some reason, the serializer is not able to unmarshal it. My guess is that is related to this: https://github.com/java-operator-sdk/java-operator-sdk/pull/311
After adding `"io.fabric8.kubernetes.api.model.Status"` it to my local reflection-config.json file the problem seems to be fixed.

**Comment by metacosm** [comment]
@ruromero which version of the SDK are you using?

**Comment by ruromero** [comment]
ATM we're using 1.7.0
Should it be fixed in 1.8.0? I've tried to upgrade but there are some breaking changes

**Comment by metacosm** [comment]
It's probably addressed in later versions, yes but that indeed means handling breaking changes.
